_posts/2016-04-02-xss-justthefacts.md

```diff
@@ -0,0 +1,48 @@
+
+Cross-site scripting is the slightly clunky name for the malicious insertion of unauthorised javascript into a website or web application. Javascript can be dangerous because it executes client-side within the context of the end user's browser and hence can have access to customer data as well as control of browser actions. Malicious payloads can include:
+* Capturing user input such as passwords via a keylogger
+* Sending cookies, tokens and other cached data to a third party
+* Performing network requests and system operations that the user hasn't requested
+* Forcing downloads of files to the end user PC
+
+At its simplest, XSS relies on a website returning executable code to a client that a third party has injected, rather than data that the webserver is intending to serve. This can take a variety of forms but usually relies on the website returning a parameter to a customer that the customer has previously entered in an earlier web request. That is, it relies on data that is intended to be passive/handled data (such as raw text) to transform context into some form of exectuable payload such as a command or script, which can then execute. If the parameter containing the executable code is passed to and executes on the server side, it can lead to a SQL injection or command execution attack, and if executed on the client side, it can lead to an XSS attack.
+
+An example of a website that could be vulnerable to this could be as simple as a site containing a single form that echoes your name back to you: you enter your name ("Bob") in a login form field, and the website returns "Hello, Bob" in its HTTP response. If the website can instead be forced to process the name of "&lt;script&amp;... &lt;script&amp;" then it is possible to inject javascript into unsuspecting visitors' browsers.
+
+![Bobby Tables cartoon](https://imgs.xkcd.com/comics/exploits_of_a_mom.png)
+
+It is this combination of the widespread susceptibly to XSS with the powerful payloads that the vulnerability that can deliver and the relative simplicity of exploiting it that make XSS such an important threat to protect against.
+
+The reason that XSS vulnerabilities are so prevalent is that whilst XSS is a trivial vulnerability to prevent against, doing so requires robust and consistent coding practices that are consistently implemented over time - the consideration of data handling of input and output data by almost every piece of code you write at every level of your stack.
+
+The correct approach to protect your website against XSS and other linked vulnerabilities is to ensure that any system the data is passed onto treats the data as raw data/text and doesn't allow the data to context-jump into an executable command. In practice this means performing some combination of:
```

Comment:
"any system the data is passed onto" is clunky... just "all input data is treated" should be sufficient